### DISCUSSION TITLE
Support constants from RFC9110

### DISCUSSION THREAD

**Discussion by mbeijen** [encode/httpx#3162] [Potential Issue]
`httpx.codes` contains constants for HTTP status codes. However it does not support the names as defined in RFC9110.

https://www.rfc-editor.org/rfc/rfc9110.html#name-changes-from-rfc-7231

This RFC for instance defines some status codes previously only used in WEBDAV and assigns more generic names to them, such as HTTP 422, which previously was `UNPROCESSABLE ENTITY` and now is `UNPROCESSABLE CONTENT`.  There are several more.

I found out about this issue because one of the systems we interface with started returning HTTP 422 Unprocessable Content, and I wanted to catch this in code, and I could not find the constant for `UNPROCESSABLE_CONTENT`, only for `UNPROCESSABLE_ENTITY`, which made me think that might be a different error, but it's just the 'old' name.

So I think httpx should also support the status codes as described and named in RFC9110. The old constants should be kept for backwards compatibility.

 'core' python also carries these status codes. They have the same issue, I've created https://github.com/python/cpython/pull/117611 for that.

**Comment by mbeijen** [comment]
My PR for core python has been merged :rocket:  but `httpx` does not support RFC9110 status code names yet :cry:

**Reply by lovelydinosaur** [reply]
Perhaps we could start by enumerating in this discussion the exact changes...

https://github.com/encode/httpx/blob/4b85e6c3898b94e686b427afd83138c87520b479/httpx/_status_codes.py#L155-L157

* What would be new on this list?
* What would be changed on this list?

**Reply by mbeijen** [reply]
There are no new status codes defined in RFC9110, but some are renamed. See https://www.rfc-editor.org/rfc/rfc9110.html#name-changes-from-rfc-7231

As I said, I found out about this issue because one of the systems we interface with started returning HTTP 422 Unprocessable Content, and I wanted to catch this in code, and I could not find the constant for UNPROCESSABLE_CONTENT, only for UNPROCESSABLE_ENTITY, which made me think that might be a different error, but it's just the 'old' name.

I propose that httpx/_status_codes uses the names for constants as defined in RFC9110. The old names also should be kept around not to break existing code. 

It could be like this, where we map the old define to the new define. Python chose to do it like this in their http status codes:

```
    UNPROCESSABLE_CONTENT = 422, "Unprocessable Content"
    ...
    UNPROCESSABLE_ENTITY = UNPROCESSABLE_CONTENT
```

Or it could be like this:

```
    UNPROCESSABLE_CONTENT = 422, "Unprocessable Content"
    UNPROCESSABLE_ENTITY = 422, "Unprocessable Entity"
```

The difference there is that the contents of `UNPROCESSABLE_ENTITY` does not change; I'm not sure if this is a benefit or not.

In total, 4 status codes have updated wording in RFC9110, see my commit here:
https://github.com/mbeijen/httpx/commit/6f500a2c1afd200a0a3011a562d4b3c6cb023067
